src/mca/state/dvm/state_dvm.c

```diff
@@ -417,6 +417,7 @@ static void ready_for_debug(int fd, short args, void *cbdata)
     if (jdata->num_procs == jdata->num_ready_for_debug) {
         /* generate the event notifying any connected tool that
          * the specified job is ready for debug */
+        // TODO?
```

Comment:
Aha! You have stepped into one of my "honey traps". 😸 Just a way for me to track people's progress on the code and how closely they are following/reading the Standard while I work on other things.

I've been watching as these example programs have all had a common mistake in them - not catastrophic and easily fixed, yet indicative that people aren't carefully reading the Standard. FWIW: any time I touch these areas and/or examples, I keep the relevant Standard chapter open on my monitor. I can't count the number of times I thought I remembered how something worked, only to catch myself with a quick recheck of the actual doc. There is a surprising amount of detailed explanation in there.

This was something I was pretty confident people would miss. The example tools are all waiting on the `PMIX_LAUNCH_COMPLETE` event - but that only tells you that the procs have all been exec'd. It _doesn't_ mean that the procs have reached the indicated point where they are to "hold" waiting for debugger attach.

The debugger needs to wait for the `PMIX_DEBUG_WAITING_FOR_NOTIFY` event instead of "launch complete" - this tells the debugger that all procs are at the designated "hold" point:
```text
The launcher (RM or IL) is responsible for generating the PMIX_DEBUG_WAITING_FOR_NOTIFY
event when all processes have reached the pause point.
```
The pause point can be on first instruction (for `PMIX_DEBUG_STOP_ON_EXEC`), in PMIx_Init (for `PMIX_DEBUG_STOP_IN_INIT`), or an arbitrary place in the process (for `PMIX_DEBUG_WAIT_FOR_NOTIFY`). 

IIRC, I had made the necessary changes to PMIx itself - the server needs to know when the procs are waiting_in_init or waiting_in_program if one of those was the directed "hold" point - and had done some of the event trapping in PRRTE. Just left this last state machine integration dangling out there as a "trap".

I'll complete the integration now that you've stepped into it. I'm gradually beginning to fill in the other "honey traps" as it is becoming clear people aren't going to find them before I depart (sadly sigh).

Reply:
Uh ok. Yeah, I noticed this and made a corresponding note to myself in the rmaps change here. I wanted to sync the current version of the examples upstream before tackling the `PMIX_DEBUG_WAITING_FOR_NOTIFY` next. I also noticed that the daemon is sending multiple 'release' messages to the application which causes it to hang - which I have a fix for.

Do you mind if I take a pass at the `PMIX_DEBUG_WAITING_FOR_NOTIFY` event in this PR?

If you want to file issues for any leftover "honey traps" we can make sure to cover them.

Reply:
Sure, feel free. I was planning on renaming it `PMIX_READY_FOR_DEBUG` (I may even have that in one of the Standard RFCs) as the debugger isn't necessarily going to issue a "release" event (it may twiddle a memory location instead). I'd suggest going that direction as it is clearer.

Note that the plan for "stop in init" and for "wait in the app" was for the PRRTE daemon to aggregate "ready" notifications from the local clients and then send a notification to the DVM master, which would aggregate across daemons before notifying the tool. This means those notifications need to have "only to local RM" range. "Stop on exec" works the same except the daemon already tells the DVM master that the proc was exec'd, so the master has all it needs to notify the tool.

Reply:
I ran out of time to tackle this issue in this PR. Since it will involve changes to OpenPMIx and PRRTE we will tackle it in another PR soon. I filed an issue so we don't forget:
 * https://github.com/openpmix/prrte/issues/885